V3.0.0 - replace python mqtt script with mqtt.js

LeLunZ opened the pull request:
- replaced python mqtt script with mqtt.js
    - checkTVStatus still needs to run periodically to check if the tv is online/offline
    - but the Accessory now opens a mqtt connection if the tv is online, and updates states in real time
- there is now a node executable used to authenticate the homebridge instance (instead of the py script) 

Most changes should already be reflacted in the README
  
TODO
- Add option for TVs with always on display
- Add apps (as source) 
- Remove duplicated code

*still work in progress* (but feedback is always welcome)

ROADMAP: 
Current beta v2.0.2 is currently being tested by me, and seems to function like normal. If thats staying the case for the coming days and I don't find any issue, the beta will be released as a normal release (as the last version of v2).
And then this will be made a beta for testing.

MrAsterisco commented:
Hey @LeLunZ ! It seems you're still working on this PR, so I converted it to a draft to avoid accidental merges. Please let me know when you'd like for me to review 😊 
Thank you for pushing this forward!

LeLunZ commented:
I want to give a update on "Always On TVs" #18 :

I read through the home-assistent thread mentioned [here](https://github.com/MrAsterisco/homebridge-hisense-tv/issues/18#issuecomment-1247593321)
And it seems there are two different ways Always On TVs can show the Sleep State:
1.  if statetype (in the state topic we already listen to) is set to 'fake_sleep_0'
2. topic .../platform_service/data/picturesetting sends out Settings object with menu_items which one could change depending on if the TV is On or Off.

To get 1. to work, a user just need to change "TV Type" in the settings to Fake Sleep.
To get 2. to work, someone really must look up what menu_item changes and add the id, and the value it has in the off state to the config.

LeLunZ commented:
From my side this is finally done ;)

@MrAsterisco if you have the time and want to take a look at the code please do.
I am not that skillful in writing README files so if you find any weird wordings there please tell me :)

I also released the first v3.0.0 beta so that this is testable while in review.

--- 

Furthermore just some information about the changes:
Changes in homekit are now mostly triggered reactive based. (Means we only do polling if 100% necessary. For every other state updates, we get mqtt messages and then update the homekit state based on that)

For normal TVs we can never fully remove polling, because they shut down their mqtt server if turned off. 
The mqtt client could do the polling itself, but I didn't change the interval (60 seconds) to something lower, as its probably more resource intensive than just opening the socket for a very short time???

LeLunZ commented:
I guess I should also add more information to the  CHANGELOG.md so users who just upgrade without looking into the repo see that there are some breaking changes.

LeLunZ commented:
Well finally done... Saw that the homebridge-plugin-template migrated to esmodules so i included that.
Also removed the comment from above to put it here again:

From my side this is finally done ;)

@MrAsterisco if you have the time and want to take a look at the code please do.
I am not that skillful in writing README files so if you find any weird wordings there please tell me :)

I also released the first v3.0.0 beta so that this is testable while in review.

--- 

Furthermore just some information about the changes:
Changes in homekit are now mostly triggered reactive based. (Means we only do polling if 100% necessary. For every other state updates, we get mqtt messages and then update the homekit state based on that)

For normal TVs we can never fully remove polling, because they shut down their mqtt server if turned off. 
The mqtt client could do the polling itself, but I didn't change the interval (60 seconds) to something lower, as its probably more resource intensive than just opening the socket for a very short time???

---

Also regarding the previously known issues:

> Switching input to "TV" might not work properly. Home will not display any error, but the next TV state refresh will bring the input back to the previous one (which is also the one displayed on the TV).

Is now fixed. That was a problem with always creating the mqtt client (when publishing), where the connection probably wasn't 100% there or stable. Now when pressing a input it always works as the mqtt connection is now always open :)

> - Making changes to the TV state (turning on/off, changing input) while the Home app is opened will not trigger a live update. *This is theoretically supported by the plugin, but it seems to not work properly.*. Just switching to another app and then going back to Home will trigger a refresh.

Is now also fixed with the migration to the newer homebridge version and the reactive approach of updating states.
